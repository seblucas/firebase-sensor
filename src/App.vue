<template>
    <div id="app" v-bind:class="[fluid ? 'container-fluid' : 'container']">
      <nav class="navbar navbar-default">
        <div class="container-fluid">
          <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#menu-1" aria-expanded="false">
              <span class="sr-only">Toggle navigation</span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
            </button>
            <router-link class="navbar-brand" :to="{ name: 'HomePage'}">Capteurs</router-link>
          </div>
          <div class="collapse navbar-collapse" id="menu-1">
            <ul class="nav navbar-nav navbar-right">
              <li><a href="#" target="_self" v-on:click="toggleHZoom()"><span class="glyphicon glyphicon-resize-horizontal"></span></a></li>
              <li><router-link :to="{ name: 'LineChartPage'}"><span class="glyphicon glyphicon-signal"></span></router-link></li>
              <li><router-link :to="{ name: 'ErrorListPage'}"><span class="badge progress-bar-danger">{{ errors.length }}</span></router-link></li>
              <li class="dropdown">
                <a href="#" class="dropdown-toggle" target="_self" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false"><span class="glyphicon glyphicon-user"></span> <span class="caret"></span></a>
                <ul class="dropdown-menu">
                  <li v-if="!authData"><a href="#" target="_self" v-on:click="$store.dispatch('login')">Log In</a></li>
                  <li v-if="authData"><a href="#" target="_self">Uid {{ authData.uid }}</a></li>
                  <li v-if="authData"><a href="#" target="_self">Logged as {{ authData.providerData[0].displayName }}</a></li>
                  <li v-if="authData"><a href="#" target="_self" v-on:click="$store.dispatch('logout')">Log Out</a></li>
                </ul>
              </li>
            </ul>
          </div>
        </div>
      </nav>
      <div class="row">
        <div class="col-xs-12">
          <div class="alert alert-warning" role="alert" v-if="!authData">
            Please Log in to use this page. <span v-if="statusLoaded">Status : {{ status }}</span>
          </div>
        </div>
      </div>
      <div v-if="rooms && errors && categories">
        <router-view/>
      </div>
      <div class="visible-xs hidden-sm hidden-md hidden-lg">xs</div>
      <div class="hidden-xs visible-sm hidden-md hidden-lg">sm</div>
      <div class="hidden-xs hidden-sm visible-md hidden-lg">md</div>
      <div class="hidden-xs hidden-sm hidden-md visible-lg">lg</div>
    </div>
</template>

<script>
import { ref, get, query, limitToLast } from 'firebase/database'

export default {
  name: 'app',
  data () {
    return {
      fluid: false,
      tstamp: 0,
      statusLoaded: false,
      statusData: {
        grenier: 0,
        palier: 0,
        deulemont: 0,
        cuisine: 0
      }
    }
  },
  computed: {
    authData () {
      return this.$store.getters.user
    },
    rooms () {
      return this.$store.getters.rooms
    },
    errors () {
      return this.$store.getters.errors
    },
    categories () {
      return this.$store.getters.categories
    },
    firebaseDatabase () {
      return this.$store.getters.firebaseDatabase
    },
    status () {
      if (this.authData) return 'Connected'
      let error = 0
      Object.keys(this.statusData).forEach((roomId) => {
        if (this.tstamp - this.statusData[roomId] > 31 * 60) error = error + 1
      })
      if (error > 0) return error + ' sensor failed'
      return 'Everything is up'
    }
  },
  methods: {
    toggleHZoom () {
      this.fluid = !this.fluid
    },
    async initData () {
      Object.keys(this.statusData).forEach(async (roomId) => {
        const lastReadings = query(ref(this.firebaseDatabase, 'readings/' + roomId), limitToLast(1))
        const newValue = await get(lastReadings)
        const myArray = this.ObjectToArray(newValue.val())
        this.statusData[roomId] = myArray[0].time
      })
      console.log(this.statusData)
    }

  },
  beforeCreate () {
    this.$store.commit('setFirebaseApp', this.$firebase)
    this.$store.dispatch('listenForAuthentication')
  },
  async mounted () {
    this.tstamp = Math.floor(Date.now() / 1000)
    await this.initData()
    this.statusLoaded = true
  }
}
</script>

<style lang="scss">
// @import 'nvd3/build/nv.d3.css';
</style>
