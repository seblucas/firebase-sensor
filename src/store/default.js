import { getAuth, onAuthStateChanged, GoogleAuthProvider, signInWithPopup, signOut } from 'firebase/auth'
import { getDatabase, ref, get, onValue, remove, query, limitToLast } from 'firebase/database'

import ObjectToArray from '@/helper/object2array'
import devLog from '@/helper/devLog'

export const state = {
  firebaseApp: false,
  firebaseAuth: false,
  firebaseDatabase: false,
  user: null,
  rooms: false,
  categories: false,
  errors: []
}

export const mutations = {
  setUser (state, payload) {
    state.user = payload
  },
  setRooms (state, payload) {
    state.rooms = payload
  },
  setCategories (state, payload) {
    state.categories = payload
  },
  setErrors (state, payload) {
    state.errors = payload
  },
  setFirebaseApp (state, payload) {
    state.firebaseApp = payload
    state.firebaseAuth = getAuth()
    state.firebaseDatabase = getDatabase(state.firebaseApp)
  }
}

export const actions = {
  listenForAuthentication ({ dispatch, commit, getters }) {
    onAuthStateChanged(getters.firebaseAuth, (authData) => {
      if (authData) {
        devLog('Logged in as:', authData.uid)
        commit('setUser', authData)
        dispatch('loadRooms')
        dispatch('loadCategories')
        dispatch('loadErrors')
      } else {
        devLog('Logged out')
        commit('setUser', null)
        commit('setRooms', null)
        commit('setCategories', null)
        commit('setErrors', [])
      }
    })
  },
  login ({ getters }) {
    const provider = new GoogleAuthProvider()
    signInWithPopup(getters.firebaseAuth, provider).then(() => {
      // No need to do anything here it's handled by $onAuthStateChanged
    }).catch((error) => {
      console.error('Authentication failed:', error)
    })
  },
  logout ({ getters }) {
    signOut(getters.firebaseAuth).then(() => {
      devLog('Unauthentication completed')
    }).catch((error) => {
      console.error('Unauthentication failed:', error)
    })
  },
  loadRooms ({ commit, getters }) {
    get(ref(getters.firebaseDatabase, 'rooms')).then((newValue) => {
      commit('setRooms', newValue.val())
    }).catch((error) => {
      console.error(error)
    })
  },
  loadCategories ({ commit, getters }) {
    get(ref(getters.firebaseDatabase, 'readingCategories')).then((newValue) => {
      commit('setCategories', newValue.val())
    }).catch((error) => {
      console.error(error)
    })
  },
  loadErrors ({ commit, getters }) {
    const errors = ref(getters.firebaseDatabase, 'errors')
    onValue(errors, (newValue) => {
      commit('setErrors', ObjectToArray(newValue.val()))
    })
  },
  async loadDataFromFirebase ({ getters }, { roomId, startTimestamp, numberOfHours, categoryId }) {
    const queryReadings = query(ref(getters.firebaseDatabase, 'readings/' + roomId), limitToLast(4 * (numberOfHours + 1)))
    const newValue = await get(queryReadings)
    devLog(`Loaded from database for ${roomId}`)
    let basicArray = ObjectToArray(newValue.val())

    basicArray = basicArray.filter((item) => {
      // Remove too old readings and readings without the category
      return item.time > startTimestamp &&
        (!categoryId || Object.prototype.hasOwnProperty.call(item, categoryId))
    })
    return basicArray
  },
  async removeError ({ getters }, item) {
    await remove(ref(getters.firebaseDatabase, 'errors/' + item.id))
  }
}

export const getters = {
  firebaseAuth (state) {
    return state.firebaseAuth
  },
  firebaseApp (state) {
    return state.firebaseApp
  },
  firebaseDatabase (state) {
    return state.firebaseDatabase
  },
  user (state) {
    return state.user
  },
  rooms (state) {
    return state.rooms
  },
  categories (state) {
    return state.categories
  },
  errors (state) {
    return state.errors
  },
  categoriesSorted (state) {
    const basicArray = ObjectToArray(state.categories)
    return basicArray.sort((a, b) => a.order - b.order)
  }
}
