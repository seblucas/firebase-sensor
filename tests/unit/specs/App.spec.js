import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Router from 'vue-router'
import App from '@/App'
import { mockStore } from '../mocks/store.mock'

describe('App.vue', () => {
  let localVue
  let store

  beforeEach(() => {
    localVue = createLocalVue()

    localVue.use(Vuex)
    localVue.use(Router)

    store = new Vuex.Store(mockStore)
  })

  it('shows a page', () => {
    const wrapper = shallowMount(App, {
      store,
      localVue
    })
    // workaround to please coverage check
    expect(wrapper.vm.categories).toBeFalsy()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('switch to full width mode if icon is clicked', async () => {
    const wrapper = shallowMount(App, {
      store,
      localVue
    })
    const mainContainer = wrapper.find('#app')
    expect(mainContainer.classes()).toContain('container')
    const zoomIcon = wrapper.find('.glyphicon-resize-horizontal')
    zoomIcon.trigger('click')
    await wrapper.vm.$nextTick()
    expect(mainContainer.classes()).toContain('container-fluid')
  })
})
