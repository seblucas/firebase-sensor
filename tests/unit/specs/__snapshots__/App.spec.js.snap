// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue shows a page 1`] = `
<div id="app" class="container">
  <nav class="navbar navbar-default">
    <div class="container-fluid">
      <div class="navbar-header"><button type="button" data-toggle="collapse" data-target="#menu-1" aria-expanded="false" class="navbar-toggle collapsed"><span class="sr-only">Toggle navigation</span> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span></button>
        <router-link-stub to="[object Object]" tag="a" ariacurrentvalue="page" event="click" class="navbar-brand">Capteurs</router-link-stub>
      </div>
      <div id="menu-1" class="collapse navbar-collapse">
        <ul class="nav navbar-nav navbar-right">
          <li><a href="#" target="_self"><span class="glyphicon glyphicon-resize-horizontal"></span></a></li>
          <li>
            <router-link-stub to="[object Object]" tag="a" ariacurrentvalue="page" event="click"><span class="glyphicon glyphicon-signal"></span></router-link-stub>
          </li>
          <li>
            <router-link-stub to="[object Object]" tag="a" ariacurrentvalue="page" event="click"><span class="badge progress-bar-danger"></span></router-link-stub>
          </li>
          <li class="dropdown"><a href="#" target="_self" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false" class="dropdown-toggle"><span class="glyphicon glyphicon-user"></span> <span class="caret"></span></a>
            <ul class="dropdown-menu">
              <li><a href="#" target="_self">Log In</a></li>
              <!---->
              <!---->
              <!---->
            </ul>
          </li>
        </ul>
      </div>
    </div>
  </nav>
  <div class="row">
    <div class="col-xs-12">
      <div role="alert" class="alert alert-warning">
        Please Log in to use this page
      </div>
    </div>
  </div>
  <!---->
  <div class="visible-xs hidden-sm hidden-md hidden-lg">xs</div>
  <div class="hidden-xs visible-sm hidden-md hidden-lg">sm</div>
  <div class="hidden-xs hidden-sm visible-md hidden-lg">md</div>
  <div class="hidden-xs hidden-sm hidden-md visible-lg">lg</div>
</div>
`;
